Fix single-quotes incorrectly escaped when dragging into script editor

anvilfolk opened the pull request:
Fixes #56263. Fundamentally the same as godotengine/godot#57085.

Opted to leave ustring.cpp alone, as suggested by rrohak, and instead add a method to ScriptTextEditor that escapes a String for drag-n-drop into the script window according to single-quote preferences. It does that by de-escaping single-quotes that were escaped by String::c_escape, only when relevant.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

anvilfolk committed:
Fix single-quotes incorrectly escaped when dragging into script editor

Chaosus commented:
We don't use sentences like "Fixes [something]" in commits and PR names, please rename it to reflect the change itself like: "Fix ' in file name gets incorrectly escaped when dragged to script editor" or shorter, don't forget to amend the commit name as well as the PR name.

anvilfolk commented:
> We don't use sentences like `Fixes [link]` in commits and PR names, please rename it to reflect the change itself like: `Fix ' in file name gets incorrectly escaped when dragged to script editor` or shorter, don't forget to amend the commit name as well as the PR name.

Done, and apologies. A lot to keep track of and messed that up. Hopefully fixed!

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+
 void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from) {
-	const String quote_style = EDITOR_GET("text_editor/completion/use_single_quotes") ? "'" : "\"";
+	const bool using_single_quotes = EDITOR_GET("text_editor/completion/use_single_quotes");
+	const String quote_style = using_single_quotes ? "'" : "\"";
```
Can we remove the need for this and put `.quote` inside  `_escape_drop_data`?

anvilfolk replied:
Yes, excellent suggestion! Changed _escape_drop_data to _quote_drop_data which already escapes and quotes as needed!

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
 }
 
+
+String ScriptTextEditor::_escape_drop_data(const String &str) {
```
Could we make this static?

anvilfolk replied:
Done! No reason it shouldn't be indeed :)

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,8 +1598,27 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+	
+	String escaped = str.c_escape();
+
+	// Resource paths, node paths and object properties cannot contain double-quotes
+	// so we do not bother reverting back the .replace from c_escape.
```
This comment seems a bit out of place.

anvilfolk replied:
Agreed, leftover from a previous understanding of the situation in which I thought filenames could have double-quotes. They cannot, so now we just don't bother worrying about them :)

Paulb23 reviewed (changes requested):
Looks good, couple small changes commented inline.

Applied the patch locally and it complained about a formatting issue would recommend setting up clang-format, can use the guide here: https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html

anvilfolk commented:
> Looks good, couple small changes commented inline.
> 
> Applied the patch locally and it complained about a formatting issue would recommend setting up clang-format, can use the guide here: https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html

Agreed and implemented all the changes, squashed commit and force-pushed. First time doing all this, but it looks like it might've worked as expected :)

I am using Visual Studio 2022 and it's supposed to apply the .clang-format file automatically (even warns me of it), but I went ahead and force-formatted the .h and .cpp file, so hopefully we won't have any issues!

Paulb23 commented on editor/plugins/script_text_editor.h:
```diff
@@ -196,6 +196,7 @@ class ScriptTextEditor : public ScriptEditorBase {
 	void drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from);
 
 	String _get_absolute_path(const String &rel_path);
+	static String _quote_drop_data(const String &str);
```
No need to put it in the header if it's static.

anvilfolk replied:
I'm trying to figure out what the best way to define this type of method:
- It's a helper method for ScriptTextEditor
- It's not used outside of ScriptTextEditor (therefore shouldn't be in the .h file like you said)
- Someone in dev chat was saying it's uncommon to have methods that aren't associated with a class.

`String _quote_drop_data(const String &str) { ... }` seems to counter that last point.
`static String ScriptTextEditor::_quote_drop_data(const String &str) { ... }` doesn't compile without it also in the .h file
`static String _quote_drop_data(const String &str) { ... }` compiles but not entirely sure what it does, hehe

Haven't done C++ in almost 2 decades, so probably missing something here :)

anvilfolk replied:
Pushed a change using the first method, which I think was your suggestion. Happy to change if it wasn't :)

Paulb23 replied:
Should be the latter one `static String _quote_drop_data(const String &str) { ... }` :)

It's from C to make the function only visible in this file.

It's not the most common pattern, but used a few times in the `ScriptEditors` for small utility methods like this, see `_find_script_node` above this one. There's a few more further up too.

anvilfolk replied:
Changes submitted!

I was assuming C++, will look into static in C, thank you!

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+	}
+
+	// Add preferred quotes.
+	const String quote_style = using_single_quotes ? "'" : "\"";
+
+	return escaped.quote(quote_style);
```
```suggestion
	return escaped.quote(using_single_quotes ? "'" : "\"");
```

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+
+	const bool using_single_quotes = EDITOR_GET("text_editor/completion/use_single_quotes");
+
+	// Escape any special characters in the string
```
Useless comment as it duplicates the line of code.

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+		escaped = escaped.replace("\\'", "\'"); 
+	}
+
+	// Add preferred quotes.
```
Ditto.

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1598,9 +1598,28 @@ static Node *_find_script_node(Node *p_edited_scene, Node *p_current_node, const
+	// Escape any special characters in the string
+	String escaped = str.c_escape();
+
+	// Unescape single quotes if user prefers double-quote delimited strings.
```
Should explain the why a bit cleaner, something like:
```suggestion
	// If string is double quoted, there is no need to escape single quotes, so we can remove the extra escaping added above here.
```

Paulb23 reviewed (approved):
LGTM

mhilbrunner commented:
Thank you for fixing this! 🎉
